src/lib.rs

```diff
@@ -126,15 +126,49 @@ pub trait Renderable: Send + Sync {
     fn render(&self, context: &mut Context) -> Result<Option<String>>;
 }
 
+pub type TemplateLocation = Path;
+pub trait FileRepository {
+    fn read_template_file(&self, path: &TemplateLocation) -> Result<String>;
```

Comment:
I think there is a valuable opportunity for caching here.

I'm just trying o think of what should be cached.  Could we make it so the `Template` gets cached?  Do we need to instead somehow expose `Vec<Element>` (ie token stream) and cache that?

Any thoughts?  If you don't mind the break, this is a bigger thing that could be deferred to later.